### PR TITLE
feat(import-service): add import service

### LINKED ISSUE #1281: Add import service
**Is your feature request related to a problem? Please describe.**
Need a microservice to import excel file to Database.

**Describe the solution you'd like**
We should have an import service which will provide feature of importing the excel file to database after downloading it from S3 and  also upload feature for the excel file.

### PR DESCRIPTION
**Pull request by Surbhi-sharma1**
add import service

GH--1281

## Description

Added Import service

Fixes #1281 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [ ] Performed a self-review of my own code
- [x] npm test passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: services/import-service/src/services/excel.service.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const batches: {rows: any[]; types: Record<string, string>}[][] = [];
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    sheets.forEach((sheet: any) => {
+      const rowOutlineLevel: number[] = [];
+
+      //find max outline level
+      let maxOutlineLevel = 0;
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      sheet['!rows']?.forEach((row: any, index: number) => {
```
Comment (akshatdubeysf) [inline services/import-service/src/services/excel.service.ts:32]: again, no any

File: services/import-service/src/services/excel.service.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      });
+
+      console.log('max outline level ', maxOutlineLevel);
+      const sheetData = XLSX.utils.sheet_to_json(sheet, {
+        raw: false,
+      });
+
+      let currentOutlineLevel = 0;
+      //make batches according to levels
+      while (currentOutlineLevel <= maxOutlineLevel) {
```
Comment (akshatdubeysf) [inline services/import-service/src/services/excel.service.ts:46]: This function could easily run of memory for large files

File: services/import-service/src/services/excel.service.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        batches[currentOutlineLevel] = batches[currentOutlineLevel] ?? [];
+        const currentLevelRowData = sheetData.filter(
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          (rowData: any, index: number) => {
```
Comment (akshatdubeysf) [inline services/import-service/src/services/excel.service.ts:50]: any 🙁

File: services/import-service/src/services/excel.service.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    const batches: {rows: any[]; types: Record<string, string>}[][] = [];
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    sheets.forEach((sheet: any) => {
+      const rowOutlineLevel: number[] = [];
+
+      //find max outline level
+      let maxOutlineLevel = 0;
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      sheet['!rows']?.forEach((row: any, index: number) => {
+        rowOutlineLevel[index] = row.level;
```
Comment (akshatdubeysf) [inline services/import-service/src/services/excel.service.ts:33]: leave a comment to explain why we need this variable

File: services/import-service/src/services/excel.service.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+      let currentOutlineLevel = 0;
+      //make batches according to levels
+      while (currentOutlineLevel <= maxOutlineLevel) {
+        batches[currentOutlineLevel] = batches[currentOutlineLevel] ?? [];
+        const currentLevelRowData = sheetData.filter(
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          (rowData: any, index: number) => {
+            //assuming first row to be header row
+            if (rowOutlineLevel[index + 1] === currentOutlineLevel) {
```
Comment (akshatdubeysf) [inline services/import-service/src/services/excel.service.ts:52]: wrong way to filter out first row

File: services/import-service/src/services/excel.service.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+              return rowData;
+            }
+          },
+        );
+        //make batches
+        for (let i = 0; i < currentLevelRowData.length; i += batchSize) {
```
Comment (akshatdubeysf) [inline services/import-service/src/services/excel.service.ts:58]: Can write a generator function to improve the quality and readability of this code

File: services/import-service/src/providers/receive-message-listener.provider.ts [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    private readonly saveData: SaveDataFn,
+  ) {}
+
+  value() {
+    return () => this.receiveInvoker();
+  }
+
+  receiveInvoker() {
+    let count = 0;
+    setInterval(() => {
```
Comment (akshatdubeysf) [inline services/import-service/src/providers/receive-message-listener.provider.ts:28]: Can improve this code

Comment (akshatdubeysf) [reply services/import-service/src/providers/receive-message-listener.provider.ts:28]: What is the purpose of this, also could be written in a better way.

File: services/import-service/src/providers/receive-message-listener.provider.ts [hunk 8]
```diff
@@ -0,0 +1,96 @@
+import {
+  DeleteMessageBatchCommand,
+  DeleteMessageBatchRequestEntry,
+  ReceiveMessageCommand,
+  SendMessageCommand,
+} from '@aws-sdk/client-sqs';
+import {BindingScope, inject, injectable, Provider} from '@loopback/core';
+import {v4 as uuidv4} from 'uuid';
+import {ImportServiceBindings} from '../keys';
+import {ackQueueUrl, client, dataQueueUrl} from './send-message.provider';
```
Comment (akshatdubeysf) [inline services/import-service/src/providers/receive-message-listener.provider.ts:10]: dataQueueUrl should be in a common place

File: services/import-service/src/providers/receive-message-listener.provider.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+      const fileId = data.Messages[0].MessageAttributes?.FileId
+        .StringValue as string;
+      const totalRows = parseInt(
+        data.Messages[0].MessageAttributes?.Count.StringValue as string,
+      );
+      data.Messages.forEach(message => {
+        const recordData: MessageData = JSON.parse(message.Body as string);
+
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        this.saveData(recordData);
```
Comment (akshatdubeysf) [inline services/import-service/src/providers/receive-message-listener.provider.ts:62]: This method is provider by the user so the parameter structure should be readable and easy to understand

File: services/import-service/src/providers/receive-message-listener.provider.ts [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+      const deleteCommandReceiptHandle: DeleteMessageBatchRequestEntry[] = [];
+      data.Messages?.forEach(message => {
+        const id = uuidv4();
+        deleteCommandReceiptHandle.push({
+          Id: `${id}`,
+          ReceiptHandle: message.ReceiptHandle,
+        });
+      });
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      client.send(
```
Comment (akshatdubeysf) [inline services/import-service/src/providers/receive-message-listener.provider.ts:77]: What if the delete fails?

File: services/import-service/src/providers/receive-message-listener.provider.ts [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        }),
+      );
+      const count = fileCount.get(fileId);
+      if (count === totalRows) {
+        const input = {
+          QueueUrl: ackQueueUrl,
+          MessageBody: fileId,
+        };
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        client.send(new SendMessageCommand(input));
```
Comment (akshatdubeysf) [inline services/import-service/src/providers/receive-message-listener.provider.ts:90]: What if this fails?

File: services/import-service/src/providers/send-message.provider.ts [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+  implements Provider<(data: MessageData[][]) => Promise<void>>
+{
+  constructor(/* Add @inject to inject parameters */) {}
+
+  value() {
+    return (levelWiseBatches: MessageData[][]) =>
+      this.sendMessageLister(levelWiseBatches);
+  }
+
+  async sendMessageLister(levelWiseBatch: MessageData[][]) {
```
Comment (akshatdubeysf) [inline services/import-service/src/providers/send-message.provider.ts:28]: spelling

File: services/import-service/src/providers/send-message.provider.ts [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+      messageGroup.forEach((message, index) => {
+        // MAKE PARAMS CONFIGURABLE
+        params.Entries.push({
+          MessageAttributes: {
+            Title: {
+              DataType: 'String',
+              StringValue: 'Import from Excel Data',
+            },
+            Author: {
+              DataType: 'String',
```
Comment (akshatdubeysf) [inline services/import-service/src/providers/send-message.provider.ts:104]: ??

File: services/import-service/src/providers/send-message.provider.ts [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+        Entries: [],
+      };
+
+      messageGroup.forEach((message, index) => {
+        // MAKE PARAMS CONFIGURABLE
+        params.Entries.push({
+          MessageAttributes: {
+            Title: {
+              DataType: 'String',
+              StringValue: 'Import from Excel Data',
```
Comment (akshatdubeysf) [inline services/import-service/src/providers/send-message.provider.ts:101]: ??

File: services/import-service/README.md [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+## Basic Use
+
+Configure and load ImportServiceComponent in the application constructor
+as shown below.
+
+```ts
+import {ImportServiceComponent, ImportServiceComponentOptions, DEFAULT_IMPORT_SERVICE_OPTIONS} from 'import-service';
+// ...
+export class MyApplication extends BootMixin(ServiceMixin(RepositoryMixin(RestApplication))) {
```
Comment (akshatdubeysf) [inline services/import-service/README.md:21]: This needs to actually explain the working and usage na?

**Review by akshatdubeysf — CHANGES_REQUESTED** [review]

File: services/import-service/README.md (hunk 15 above)
Comment (Surbhi-sharma1) [reply services/import-service/README.md:21]: ok Sir 
will do all the changes

File: services/import-service/src/controllers/file-upload.controller.ts [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    private readonly multerS3Provider: IUploader,
+    @service(FileHelperService)
+    private readonly fileHelperService: FileHelperService,
+  ) {}
+
+  // @authenticate(STRATEGY.BEARER, {
+  //   passReqToCallback: true,
+  // })
+  // @authorize({permissions: ['*']})
+  @post('/fileupload', {
```
[the remark this answers is not in the document]
Comment (Surbhi-sharma1) [reply services/import-service/src/controllers/file-upload.controller.ts:29]: done.

File: services/import-service/src/controllers/file-upload.controller.ts [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+      },
+    })
+    request: Request,
+    @inject(RestBindings.Http.RESPONSE) response: Response,
+    @inject(AWSS3Bindings.AwsS3Provider) s3: AWS.S3,
+  ): Promise<AnyObject> {
+    const safeMulterS3Options: SafeMulterS3Options = {
+      s3,
+      bucket,
+      // Set public re,ad permissions
```
[the remark this answers is not in the document]
Comment (Surbhi-sharma1) [reply services/import-service/src/controllers/file-upload.controller.ts:54]: done

File: services/import-service/src/controllers/file-upload.controller.ts [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+      // acl: 'public-read',
+      //Set key/ filename as original uploaded name
+      key: (_req, file, cb) => {
+        const fileSplitArr = file.originalname.split('.');
+        const fileExt = fileSplitArr[fileSplitArr.length - 1];
+        const fileName = fileSplitArr.splice(-1, 1).join('_');
```
[the remark this answers is not in the document]
Comment (Surbhi-sharma1) [reply services/import-service/src/controllers/file-upload.controller.ts:60]: done

File: services/import-service/src/providers/receive-message-listener.provider.ts (hunk 8 above)
Comment (Surbhi-sharma1) [reply services/import-service/src/providers/receive-message-listener.provider.ts:10]: done

File: services/import-service/src/providers/send-message.provider.ts (hunk 12 above)
Comment (Surbhi-sharma1) [reply services/import-service/src/providers/send-message.provider.ts:28]: corrected

File: services/import-service/src/services/excel.service.ts (hunk 6 above)
Comment (Surbhi-sharma1) [reply services/import-service/src/services/excel.service.ts:58]: done

**Pull request closed without merging**